File: api/schemas/component.py
from datetime import datetime

from pydantic import BaseModel


class ComponentBase(BaseModel):
    category: str
    subcategory: str | None = None
    value: str
    value_numeric: float | None = None
    value_unit: str | None = None
    package: str | None = None
    description: str | None = None
    manufacturer: str | None = None
    mpn: str | None = None
    quantity: int = 0
    min_quantity: int = 0
    location: str | None = None
    notes: str | None = None


class ComponentCreate(ComponentBase):
    pass


class ComponentUpdate(BaseModel):
    category: str | None = None
    subcategory: str | None = None
    value: str | None = None
    value_numeric: float | None = None
    value_unit: str | None = None
    package: str | None = None
    description: str | None = None
    manufacturer: str | None = None
    mpn: str | None = None
    quantity: int | None = None
    min_quantity: int | None = None
    location: str | None = None
    notes: str | None = None


class ComponentRead(ComponentBase):
    id: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class QuantityAdjust(BaseModel):
    delta: int
    reason: str = "adjustment"
    note: str | None = None


class InventoryTransactionRead(BaseModel):
    id: int
    component_id: int
    delta: int
    reason: str
    build_id: int | None
    note: str | None
    created_at: datetime

    model_config = {"from_attributes": True}

File: native/Sources/WorkshopCore/Models/Component.swift
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 Pedal Workshop Contributors
// https://github.com/ArdChoille637/pedal-workshop

import Foundation

public struct Component: Codable, Identifiable, Hashable, Sendable {
    public let id: Int
    public var category: String
    public var subcategory: String?
    public var value: String
    public var valueNumeric: Double?
    public var valueUnit: String?
    public var package: String?
    public var description: String?
    public var manufacturer: String?
    public var mpn: String?
    public var quantity: Int
    public var minQuantity: Int
    public var location: String?
    public var notes: String?
    public var createdAt: String
    public var updatedAt: String

    enum CodingKeys: String, CodingKey {
        case id, category, subcategory, value, package, description,
             manufacturer, mpn, quantity, notes, location
        case valueNumeric  = "value_numeric"
        case valueUnit     = "value_unit"
        case minQuantity   = "min_quantity"
        case createdAt     = "created_at"
        case updatedAt     = "updated_at"
    }

    public var isLowStock: Bool { minQuantity > 0 && quantity < minQuantity }
}

public struct QuantityAdjust: Codable, Sendable {
    public let delta: Int
    public let reason: String
    public let note: String?
    public init(delta: Int, reason: String = "adjustment", note: String? = nil) {
        self.delta = delta; self.reason = reason; self.note = note
    }
}
